### DISCUSSION TITLE
Sneakernet support

### DISCUSSION THREAD

**Discussion by mofosyne** [markqvist/Reticulum#470] [Ideas]
It be interesting if messages can be passed between node via flash drive or sd card

**Comment by sparkes** [comment]
Oddly, my first idea on seeing the project was using QR codes to pass messages via the most improbable routes possible.

**Reply by markqvist** [reply]
This already works :) LXMF fully supports "paper messages", which is encrypted messages passed via QR codes or **lxm://** URIs. This functionality is already available in both Sideband and Nomadnet, and any other LXMF client can implement it easily using the available API functions in the LXMF library.

**Reply by iacore** [reply]
There is a way to send 2mbps of data from video alone.
https://unnamed.website/posts/extreme-epilepsend/

**Comment by markqvist** [comment]
Full sneakernet support for the LXMF protocol is on the roadmap, but in a limited way, it's already possible:

- Set up an LXMF propagation node on a portable device, make sure auto-peering is enabled
- Have it automatically connect to RNS-enabled WiFi networks at various locations.
- Start out at location A, when the portable node announces there, any local LXMF propagation nodes will sync with it within a reasonable amount of time. If the propagation node is running within Nomadnet, the sync can be initiated manually from the TUI of nomadnet. The portable node now has messages from A on its local storage
- Move to another location, that is physically disconnected from A. The portable node automatically connects to WiFi there (or whatever else you'd want to use), and will initiate LXMF syncs with any available propagation nodes it knows there.

There's a lot of things that can be made easier and streamlined about this, and a full implementation is coming at some point (when I have the time and funds to do it), but at least it works in a somewhat limited fashion already.

**Reply by mofosyne** [reply]
You may still want to support the lower tech option of a paper sneakernet but your above idea is convenient if set up well.

However you can argue that there might be a very narrow case of paper or usb sneakernet that may still be very handy to keep in technology denied area (e.g. disasters or goverment repression). In this case, the lower tech option should be preferred.

Personally I just want to see it since it be cool, but I don't think we should overthink it as this is not likely a common use case. Just keep it simple and easy to maintain. So paper QR code would be preferred. (Btw... I'm hoping for a common standard for leaving 'QR code' dead drop for geocaching purpose... so if anyone is interested... I would love to figure out a reliable common standard for encoding a simple website into paper)

**Reply by faragher** [reply]
Rule of cool aside, paper messages are not the greatest ways of getting past people looking to stop you. Rather than getting a document, concentrating on it, and processing it, it's much lower profile to walk into a diner, have a meal, and leave, hundreds of megs heavier due to wireless transfer from a local node or another patron. 

But it is cool. Since a QR code (or the PDF417, my personal favorite) is simply data, all it needs is a processor. It provides links with a http:// prefix and LXMF paper messages are lxm:// prefixed, then the data (IIRC).

So what you'd need is a prefix, a program that handles it, and the data. I'd suggest not sending full HTML, but maybe a compressed archive or just fields to be reconstructed, but you could probably get something up and running pretty quick with the right specification.

A maximum size of 3k is pretty low by modern bloated HTML standards, but you could get a lot of text, data, and even an image or two in there if you really tried.

**Reply by KenSamson** [reply]
> Full sneakernet support for the LXMF protocol is on the roadmap, but in a limited way, it's already possible:
> 
> * Set up an LXMF propagation node on a portable device, make sure auto-peering is enabled
> * Have it automatically connect to RNS-enabled WiFi networks at various locations.
> * Start out at location A, when the portable node announces there, any local LXMF propagation nodes will sync with it within a reasonable amount of time. If the propagation node is running within Nomadnet, the sync can be initiated manually from the TUI of nomadnet. The portable node now has messages from A on its local storage
> * Move to another location, that is physically disconnected from A. The portable node automatically connects to WiFi there (or whatever else you'd want to use), and will initiate LXMF syncs with any available propagation nodes it knows there.
> 
> There's a lot of things that can be made easier and streamlined about this, and a full implementation is coming at some point (when I have the time and funds to do it), but at least it works in a somewhat limited fashion already.

Long note full of food for thought.... it's possible some of these ideas are too complex to make reliable.    

In one idea, one would be able to list in the config file of a fixed station a list of traveling LXMF nodes used for sneaker net.   I could also list in the traveling node the list of fixed nodes that I want to auto sync with.  (to create a closed network)   Once my station sees that traveling node announce, automatically initiate a sync.     

So in that case all you would really have to do is set the traveling node to announce fairly often (perhaps it could announce only over the RNODE interfaces very often) and then any stations that hear it could reply.  If you want to make it super classy, read the local telemetry and have in the traveling config file GPS coordinates, and when I am xyz close to those coordinates, then you know to announce.    

The next challenge is one I see where if 50 stations are in a Highrise complex and I pull up an announce, how do we make order out of the chaos of 50 nodes wanting to reply at the exact same time?    Well, if I listed a fixed node address, then the coordinates, my traveling node could call out to each of the fixed nodes by "name" with a sync request specific to that fixed node, meaning there is zero chance of creating a traffic jam, and I'm not driving all over town spewing announces.    I suppose we could use a button the screen or a physical button to initiate an announce as well.   

It is possible that someone might want to be able to transport and sync LXMF 100% anonymously and randomly, and maybe even enable a traveling to traveling sync as well.  (two friends have lunch in a central town once a week and exchange messages between populations of people in each of their respective towns)

If I were doing this on a raspberry pi mobile node, I would want to fire off a routine when I'm about to initiate a sync, fire off another routine when I get a reply and are working the sync, and then fire off something when I'm done.    In my brain I would use the physical IO header in the PI to turn on indicator lights of these stages, so I can run it headless and still know all is well with the location and sync.   I would know to sit tight until I got all of the messages from that location, just by looking at the lights.

This idea of off grid and remote, but still having the ability to queue messages is *much* more common in other countries like Africa, though it could really still be useful in areas where the infrastructure has been damaged, like a wide-spread fire in LA.

**Reply by xgpt** [reply]
Has any more discussion occurred elsewhere about the sneakernet aspect of this? I was under the impression that sneakernetting was more mature. It's a huge perk of the system and one that I think could lend itself to widescale adoption of the system. Specifically because it's hugely censorship resistant. Imagine, plugging in a RPI to an ethernet network that's got prop nodes that have been listening and collecting messages all over, then plugging that same RPI to another network elsewhere to share prop node messages. The lunch date from two friends between towns point is very realistic. It'd be the two nerds from each camp/city meeting up to sync up, and then they'd go back and sync up their communities. Entirely encrypted too.

**Reply by markqvist** [reply]
From an answer given elsewhere:

### On sneakernet LXMF
This is all possible right now, and was one of the class of scenarios LXMF was intended to handle.

Here's how:

- Set up two separate devices at a central place at the gathering.
    - Device A hosts an open WiFi AP
    - Device B can just be connected to A over an ethernet switch, or even just as a client on the local WiFi network
    - Anyone at the gathering can connect to the open WiFi network
- Both device A and B hosts an LXMF propagation node
- Before the event starts:
    - Make sure that device A and B propagation nodes are peered with each other.
    - Make sure that device B is peered with one or more remote propagation nodes, either on the internet or something else.
- Device A stays put at the event, all the time. Everyone can come and deposit messages on this device's propagation node.
- When both device A and B are present at the event, propagation nodes on them will sync messages between them.
- All messages from participants will get propagated to B.
- At regular intervals, someone helping facilitate the event grabs node B, walks down the mountain, to somewhere there's internet connectivity and does the following:
    - Connects device B to a WiFi network or something else
    - Sends an announce from the LXMF propagation node running on the device.
        - Other peered PNs on the wider network will now see that device B has re-appeared, and shortly begin syncing messages to it
    - Triggers a PN sync from device B to one or more peers on the wider network. All waiting messages are now distributed onto the propagation network, and PNs on that network will take care of further propagation
    - If a potential device C has already been set up at the location with internet access, that device will already have collected all relevant messages in advance, and one could manually trigger a sync from device C to device B, thus quickly getting all message onto it. This is optional of course, but would greatly speed up the process and efficiency.
- Device B is now carried back to the event, reconnected to the same network as device A
- A sync from B to A is triggered, or you just wait a few minutes for B to detect that A is now available again.
- All messages from the wider world are now available on A, and people at the event can do an LXMF sync with A to receive their messages.

I know that's all written out quite verbosely, and as a result may sound a bit convoluted, but in reality it's very simple.

Hope that helps!

All of the above can be achieved with two Raspberry Pis. Add a third one for the optional quicker full message sync. Still very doable.
   
The approach can handle _a lot_ of messages. Delivery latency may be high, but the available bandwidth is huge, and can easily get man thousands of messages in and out of an area with ease.

If you host the propagation nodes using NomadNet, you can manually control syncs with peered nodes by going to the network page and hitting Ctrl-P. Select a peered node in the list, and hit Ctrl-R to sync. May be worth setting up the relevant PNs of devices A, B and C as static peers here.

Bonus: It also provides a central message drop for people _within_ the event to communicate with each other, without having to all have constant connectivity / radios / something else. Distribute a few more WiFi APs around the event area, all connected to the same underlying LAN, and people don't even have to get off their blankets to communicate :)

There's many permutations of the same approach. If you can live with the PN not being available while it's being moved "down the mountain" you can get away with using just one. 

Or, you could replace the moving node with somebodys phone, that was just synced manually with A before heading down the mountain.

Or, you could run everything on two Android phones set up in advance.

An option of doing a manual "file export" sync would also be nice, to just move everything on a USB stick, that could be ingested at any other remote node. It's pretty easy to add, but not something I have implemented yet.

**Comment by KenSamson** [comment]
If the to: address was a "channel" name, like r:/jokes, then this sync method could then be used to send to everyone who subscribes to the r:/jokes channel, and bingo, we have "group" messaging of sorts.    Too many ideas coming along way too quickly to get them all down, but I figured I would share some of the more useful ones.

**Reply by faragher** [reply]
This is an application level idea. Group messaging is available, but it's all single point e2ee, so traffic is relatively high. Multicast with a shared key and distributed routing is the big hurdle for "group" destinations.

If you're interested in a channel-like system with a server-client mentality, I mocked up a proof of concept some time ago: https://github.com/faragher/LXMEvent

There are group chats/boards/hubs already in place, but none are truly e2ee and require multiple packets with the same payload just encrypted differently, as well as trust of the central server. I'm uncertain what the status of any of these are, but they're all linked to somewhere in this repo.
